### PR TITLE
[WIP] Conda package

### PR DESCRIPTION
**Pull request by jonmmease**
Work towards adding a conda build of orca to the CI build.

Discussion: https://github.com/plotly/orca/issues/107
Standalone prototype: https://github.com/jonmmease/orca-feedstock

### CODE REVIEW

**Comment by jonmmease** [comment]
@etpinard it doesn't look like the Travis build ran. Is there anything else I need to do?

**Comment by etpinard** [comment]
> it doesn't look like the Travis build ran (I only see AppVeyor)

Ha. It might only run on branches based off the repo origin, and not on branches based off forks. 

I should get an invite giving you write rights to this repo. I think the easier way forward would be for you to pull down this repo, cherry-pick your commits unto a new branch and make a new PR. Sorry for the inconvenience.

**Comment by etpinard** [comment]
Actually, pushing another commit to this branch should now do the trick: 

image

image

**Comment by jonmmease** [comment]
Cool, I'll give that a try first

**Comment by jonmmease** [comment]
Looks like they're all running now. Thanks!

**Comment by jonmmease** [comment]
@etpinard the conda builds for Circle/Travis/AppVeyor are all working now. Each one successfully performs the conda build and lists the directory where the build artifact is output.

The `docker-build-and-push` job is failing on "Push Docker image to Quay", which is probably a good thing since I'm not intending up push a new docker image 🙂 

All that's left now is to figure out how we want to upload the packages to anaconda cloud.  I don't have a good sense of how this part should work.  Should the CI jobs do the upload, or should we try to get the built packages out of the CI systems and upload them manually?

Also, someone with some experience with these CI tools should probably take a look at what I did to get things working 🙂

File: .circleci/config.yml [hunk 1]
```diff
@@ -21,6 +21,7 @@ workflows:
           requires:
             - test-node-v6
             - test-node-v8
+      - conda-build-and-upload
```
Comment (etpinard) [inline .circleci/config.yml]: We probably only want to run this if `test-node-v6` and `test-node-v8` succeed. Adding 

```
  requires:
     - test-node-v6
     - test-node-v8
```

should do the trick

File: .circleci/config.yml [hunk 2]
```diff
@@ -156,3 +157,29 @@ jobs:
...
+      - save_cache:
+          key: v1-source-{{ .Branch }}-{{ .Revision }}
+          paths:
+            - ".git"
+      - restore_cache:
+          key: v3-npm-{{ checksum "package-lock.json" }}
+          keys:
+            - v3-npm-
+
+      - run: wget http://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh
```
Comment (etpinard) [inline .circleci/config.yml]: We could probably cache this, but `conda-build-and-upload` runs in under 3 minutes, so I don't see the need :ok_hand:

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "orca",
-  "version": "1.0.0",
+  "version": "1.0.1-rc.1",
```
Comment (etpinard) [inline package.json]: Did you base your branch off `master` or #103 ? If the former, we should stick to `1.0.0` here.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{% set data = load_setup_py_data() %}
```
Comment (etpinard) [inline recipe/meta.yaml]: We should add a README.md file in `recipe/` briefly explanation the conda-build process.

File: .circleci/config.yml (hunk 1 above)
Comment (etpinard) [reply .circleci/config.yml]: ... and we should probably rename this step `conda-build` as we won't be uploading the builds on CI.

**Comment by jonmmease** [comment]
@etpinard 
Well that took more tries that I thought it would 🙂 The artifact zip files look good for CircleCI and AppVeyor. I don't have access to the Travis artifacts (at least I don't have an AWS account), could you check that one?

I think I also addressed all of your feedback. Let me know if the Travis artifacts need any work, and if there are any other changes you'd like to see before merging this in.

Thanks!

**Comment by etpinard** [comment]
> I don't have access to the Travis artifacts

Actually, looks like the artifacts didn't get pushed to S3 after your commits, perhaps due to another permissions restriction.

Oh well, I fetched your branch and created https://github.com/plotly/orca/compare/conda-build and all CI builds there (including the `docker-build-and-push` job on Circle) are :white_check_mark: 

Here's the link to Mac conda build:

https://s3.amazonaws.com/image-exporter-travis-artifacts/plotly/orca/433/433.1/conda-osx-64.zip

if you like to try it out.

File: recipe/README.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+Finally, build the package from the root project directory:
+
+```bash
+$ conda build recipe/
+```
+
+The resulting package will be named `plotly-orca-*.tar.bz2`, and the build
+command will display the full path of the output location. 
+
+How it works
```
Comment (etpinard) [inline recipe/README.md]: Wow. I didn't expect this much detail. Thanks for your efforts!

File: README.md [hunk 6]
```diff
@@ -14,6 +14,17 @@ Follow [@plotlygraphs](https://twitter.com/plotlygraphs) on Twitter for Orca ann
 
 ## Installation
 
+### Method 1: conda
```
Comment (etpinard) [inline README.md]: Beautifully done. Thanks!

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -114,6 +114,19 @@ automate this process somehow.
   + Grab the Mac build automatically pushed to [Amazon S3](https://s3.console.aws.amazon.com/s3/buckets/image-exporter-travis-artifacts/plotly/orca/?region=us-east-1&tab=overview) from Travis.
     **N.B.** Select the latest build (largest number *note:* the folders are not necessarily sequential) -> release.zip
 - Run `npm publish`
+- Publish conda packages:
```
Comment (etpinard) [inline CONTRIBUTING.md]: Very clear. :books:

File: recipe/README.md (hunk 5 above)
Comment (jonmmease) [reply recipe/README.md]: haha, sure. Once I got started a bunch of details came to mind that I thought would be useful background for future contributors.

**Comment by jonmmease** [comment]
Replaced by https://github.com/plotly/orca/pull/113

**Pull request closed without merging**